internal/plugin/sdk/host/interface.go

```diff
@@ -0,0 +1,83 @@
+// Package host provides the interfaces and functionality for host
+// plugins. Host plugins provide functionality for dynamic host
+// catalogs and host sets.
+package host
+
+import (
+	"context"
+
+	"github.com/hashicorp/boundary/internal/plugin/sdk/proto"
+)
+
+// HostPlugin describes the interface for host plugins.
+type HostPlugin interface {
```

Comment:
Can you explain the purpose of this interface? This seems like it's an analogue to the grpc service interface that will be generated, but it's using different parameter styles (e.g. for a service, the signature for `OnCreateCatalog` would be something like `OnCreateCatalog(ctx context.Context, req *pb.OnCreateCatalogRequest) (*pb.OnCreateCatalogResponse, error)`). Shouldn't the service be created, and its methods are the interface? Otherwise you have to translate between that one and this one.

Reply:
IMO it is good to hide that we are using grpc, and possibly even protos at all, from the plugin author.  Adding a layer between what they code against and how we are communicating with the plugin process gives us flexibility to change in the future.

Regarding what to expose in the interface, I also think the it should be the minimal one needed to easily do what they need to do.  This is especially true when we are just starting out and don't have a use case for a lot of these fiels (update time, allowed actions, boundary resource version, etc...)

Reply:
Protos simply compile to Go structs; grpc services turn into Go interfaces. I think the question is whether these structs and interfaces are sufficient for the task at hand. If we wanted to switch off grpc, would we also switch the interface? If not, why not just use the one it generates as a starting point?

At a broader level, why is it bad for plugin authors to know that it's a grpc-generated service or proto-generated structs?

Reply:
FWIW -- I don't think there is an explicit need to hide that it's grpc or proto messages; I think an interface _just to do that_ is unnecessary. I do think there is value in an interface that exists to provide convenience to the user, e.g. taking in two string parameters that under the hood go into two values in a `XRequest` object instead of creating that object yourself, as part of a plugin SDK. But, given that we can change the implementation at any time in the near term because it's only our plugins to start with and we're not doing versioning, I don't think this is _necessary_ at the moment. Fine to have, but if it forces us to define a Go API right now that we otherwise don't need to define, it might not be worth the time.

As for the proto messages -- if the intention is to make the types easier for Go developers to work with, I would argue that an interface like this one could transform those into our `api` Go structs. So we can reuse the same proto messages across the plugin interface, then parse into the API struct if we wish. But I'm not really understanding why we need to define a third version of the same types we already have in both proto and Go form, along with all the tooling we have to convert between them.

Reply:
For the proto message, the state we pass on is captured mid transaction and inadvertently exposes implementation detail.  For example, version, update time, create time, all will have different values depending on where in repository we make a call to the plugin.  It's also difficult to imagine how those fields would be helpful to the plugin, so they would just be added mental load for no benefit.